# **ASSOCIATION RULES**

## **Data Preprocessing**

In [77]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [78]:
df = pd.read_excel('/content/Online retail.xlsx', header=None)
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


0
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

In [79]:
df = df.rename(columns={0: 'Transactions'})
df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Transactions
0     shrimp,almonds,avocado,vegetables mix,green gr...
1                                burgers,meatballs,eggs
2                                               chutney
3                                        turkey,avocado
4     mineral water,milk,energy bar,whole wheat rice...
...                                                 ...
7496                      butter,light mayo,fresh bread
7497  burgers,frozen vegetables,eggs,french fries,ma...
7498                                            chicken
7499                                 escalope,green tea
7500    eggs,frozen smoothie,yogurt cake,low fat yogurt

[7501 rows x 1 columns]

## **Association Rule Mining**

### **Apply Apriori Algorithm**

In [80]:
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [81]:
df_encoded = df['Transactions'].str.get_dummies(sep=',')
df_encoded

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


asparagus  almonds  antioxydant juice  asparagus  avocado  babies food  \
0              0        1                  1          0        1            0   
1              0        0                  0          0        0            0   
2              0        0                  0          0        0            0   
3              0        0                  0          0        1            0   
4              0        0                  0          0        0            0   
...          ...      ...                ...        ...      ...          ...   
7496           0        0                  0          0        0            0   
7497           0        0                  0          0        0            0   
7498           0        0                  0          0        0            0   
7499           0        0                  0          0        0            0   
7500           0        0                  0          0        0            0   

      bacon  barbecue sauce  black tea  blueberries  ...  turkey  \
0         0               0          0            0  ...       0   
1         0               0          0            0  ...       0   
2         0               0          0            0  ...       0   
3         0               0          0            0  ...       1   
4         0               0          0            0  ...       0   
...     ...             ...        ...          ...  ...     ...   
7496      0               0          0            0  ...       0   
7497      0               0          0            0  ...       0   
7498      0               0          0            0  ...       0   
7499      0               0          0            0  ...       0   
7500      0               0          0            0  ...       0   

      vegetables mix  water spray  white wine  whole weat flour  \
0                  1            0           0                 1   
1                  0            0           0                 0   
2                  0            0           0                 0   
3                  0            0           0                 0   
4                  0            0           0                 0   
...              ...          ...         ...               ...   
7496               0            0           0                 0   
7497               0            0           0                 0   
7498               0            0           0                 0   
7499               0            0           0                 0   
7500               0            0           0                 0   

      whole wheat pasta  whole wheat rice  yams  yogurt cake  zucchini  
0                     0                 0     1            0         0  
1                     0                 0     0            0         0  
2                     0                 0     0            0         0  
3                     0                 0     0            0         0  
4                     0                 1     0            0         0  
...                 ...               ...   ...          ...       ...  
7496                  0                 0     0            0         0  
7497                  0                 0     0            0         0  
7498                  0                 0     0            0         0  
7499                  0                 0     0            0         0  
7500                  0                 0     0            1         0  

[7501 rows x 120 columns]

In [82]:
# Apply Apriori algorithm to find frequent itemsets
freq_items = apriori(df_encoded, min_support=0.02, use_colnames=True)

# Display frequent itemsets
print(freq_items)

      support                           itemsets
0    0.020397                          (almonds)
1    0.033329                          (avocado)
2    0.033729                         (brownies)
3    0.087188                          (burgers)
4    0.030129                           (butter)
..        ...                                ...
98   0.020131  (mineral water, whole wheat rice)
99   0.022930             (olive oil, spaghetti)
100  0.025197              (spaghetti, pancakes)
101  0.021197                (spaghetti, shrimp)
102  0.020931              (spaghetti, tomatoes)

[103 rows x 2 columns]


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


### **Generate Association Rules**

In [83]:
# Generate association rules
rules = association_rules(freq_items, metric="lift", min_threshold=1.0)

# Display the association rules sorted by lift
print(rules.sort_values(by='lift', ascending=False).head())


      antecedents      consequents  antecedent support  consequent support  \
64    (spaghetti)    (ground beef)            0.174110            0.098254   
65  (ground beef)      (spaghetti)            0.098254            0.174110   
86    (olive oil)      (spaghetti)            0.065858            0.174110   
87    (spaghetti)      (olive oil)            0.174110            0.065858   
79         (soup)  (mineral water)            0.050527            0.238368   

     support  confidence      lift  leverage  conviction  zhangs_metric  
64  0.039195    0.225115  2.291162  0.022088    1.163716       0.682343  
65  0.039195    0.398915  2.291162  0.022088    1.373997       0.624943  
86  0.022930    0.348178  1.999758  0.011464    1.267048       0.535186  
87  0.022930    0.131700  1.999758  0.011464    1.075829       0.605334  
79  0.023064    0.456464  1.914955  0.011020    1.401255       0.503221  


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [84]:
print(rules.head())

       antecedents     consequents  antecedent support  consequent support  \
0        (burgers)          (eggs)            0.087188            0.179709   
1           (eggs)       (burgers)            0.179709            0.087188   
2        (burgers)  (french fries)            0.087188            0.170911   
3   (french fries)       (burgers)            0.170911            0.087188   
4  (mineral water)       (burgers)            0.238368            0.087188   

    support  confidence      lift  leverage  conviction  zhangs_metric  
0  0.028796    0.330275  1.837830  0.013128    1.224818       0.499424  
1  0.028796    0.160237  1.837830  0.013128    1.086988       0.555754  
2  0.021997    0.252294  1.476173  0.007096    1.108844       0.353384  
3  0.021997    0.128705  1.476173  0.007096    1.047650       0.389069  
4  0.024397    0.102349  1.173883  0.003614    1.016889       0.194486  


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


## **Analysis and Interpretation**

###**Interpretation**:

**1. Support**: Indicates how frequently the antecedent and consequent appear together in the dataset.

*For example:*

*The rule (spaghetti) -> (ground beef) has a support of 0.039195, meaning these items appear together in approximately 3.92% of all transactions.*

**2. Confidence**: Measures how often the rule is true.

*For instance:*

*(ground beef) -> (spaghetti) has a confidence of 0.398915, indicating that 39.89% of transactions containing ground beef also contain spaghetti.*

**3. Lift**: Indicates the strength of the rule, with values greater than 1 suggesting a higher likelihood of the items being bought together than if they were independent:

*A lift of 2.291162 for (spaghetti) -> (ground beef) suggests that spaghetti and ground beef are 2.29 times more likely to be bought together than if their occurrence were independent.*

**{burgers} -> {french fries}**

**Support**: *0.0220*

**Confidence**: *0.2523*

**Lift**: *1.4762*

**Interpretation**: *Burgers and french fries are also positively associated, with a moderate lift indicating that customers buying burgers are likely to also buy french fries, though not as strongly as with eggs.*

### **Insights into Customer Behavior**

**1. Purchasing Patterns:**
Customers frequently purchase spaghetti and ground beef together, indicating they are likely used in similar recipes or meals.

**2. Cross-Selling Opportunities:**
Items like soup and mineral water have a strong association (confidence of 45.64%), suggesting that customers who buy soup are highly likely to buy mineral water as well.

**3. Product Placement:**
Knowing these associations can inform product placement strategies in stores or online platforms to increase sales of related items.

**Cross-Selling Opportunities**:
   
*Promoting eggs alongside burgers could potentially increase sales due to their high association.*

**Menu Optimization**:

*Ensure that items like burgers and french fries are often purchased together can inform menu combinations.*

**Customer Segmentation**:

*Identifying clusters of customers based on frequent itemsets can aid in targeted marketing or personalized recommendations. For instance, customers who frequently buy burgers might also be interested in related items like eggs or french fries.*

**Promotional Strategies**:

*Tailoring promotions or discounts based on these associations can potentially increase basket size or encourage repeat purchases among customers who tend to buy related items.*

## **Interview Questions**

### **1. What is lift and why is it important in Association rules?**

**Lift**: Measures how much more often items A and B occur together than expected if they were independent.  
**Importance**: Helps identify meaningful associations between items by filtering out rules where items are independent or occur together by chance.

### **2. What is support and confidence? How do you calculate them?**

**Support**: Measures the frequency of occurrence of an itemset in the dataset.   
**Calculation**: Number of transactions containing the itemset divided by total transactions.   
**Confidence**: Measures the reliability or strength of a rule.   
**Calculation**: Support of itemset A and B divided by support of itemset A.

### **3. What are some limitations or challenges of Association rules mining?**

**High Dimensionality**: Increases computational complexity and rule generation.  
**Sparse Data**: Can lead to infrequent itemsets and less meaningful associations.  
**Parameter Selection**: Choosing appropriate thresholds for support, confidence, and lift affects rule quality.  
**Interpretation**: Rules may require domain knowledge for actionable insights.   
**Overfitting**: Mining large datasets without validation may result in non-generalizable rules.